### ISSUE TITLE
Refactor Data Model - Master Skill Definitions

### ISSUE THREAD

**Issue by sven-n** [MUnique/OpenMU#49]
## What's the problem?
It seems like the current data model for master skills is too complicated. It was built under the assumption that the same skill can be a master skill at a different rank for different character classes. This was the case in previous seasons when I probably wrote the code. However, now that's not required anymore, so we can simplify things. I verified this by running a simple tool which compares all data of the same skill ids. As a source I used [this](https://raw.githubusercontent.com/TitansTech/GameServer/master/ServerConfigs/Docs%20%26%20Tools/S6%20%26%20S5E4/S6E3/SCFData/Skills/SCF_SkillTreeData.txt) file. As you can see there, there are common master skills for all characters. When you have a look at the master tree in-game, it's the left tree which is the same for every character. All other skills which are reused (e.g. Duel Master uses skills of knights and wizards), have the same rank and values.

As a side note, Webzen changed how master skills are defined in the newer seasons, too. Back then, when the master skills were introduced, every level of a master skill was defined as a separate Skill with its own ID. In Season 6, all levels of a master skill are defined under one Skill.

## So, what can be simplified?
 * A Skill doesn't need a collection of MasterSkillDefinitions. One reference to a MasterSkillDefinition is enough.
 * MasterSkillDefinition doesn't need the property for the required CharacterClass anymore, the skill itself already defines which characters are allowed to learn or use it.
 * Therefore, we don't need different MasterSkillRoots per CharacterClass, too. Three common roots for all classes are enough.
 * This simplifies the code which use the changed/removed properties - unit tests, too.

## Special case
Disclaimer: The solution for this special case is explicitly not a required part for the solution of this issue. I just want to write my thoughts about it, so that it's not forgotten.

There seems to be one skill which is different to all other skills. The "Triple Shot Mastery" requires all 10 of 10 possible master points to be allocated until it has effect. To understand my idea, I need to explain how the triple shot skill works. Currently it's not yet explicitly implemented, but handled by the common skill hit handler for skills of type SkillType.AreaSkillExplicitHits. Skills of this type send an animation packet (with a counter) and separate hit packets (also with the corresponding counter value of the animation packet). The check for the counter value is currently not yet implemented, so it's currently vulnerable to cheating. To check that there happens no cheating, we need to implement an explicit area skill hit packet handler for this particular skill (and probably for other skills as well, that's why it's way out of scope of this issue).
I see two ways to implement it here:

a) It could be implemented by adding a BaseAttributeValue with an attribute like "Triple Skill Number of Arrows" for all Elfs classes, with an initial value of 3. Each master level of the master skill adds a value of 0.1 (by Skill.PassivePowerUps). So it will reach the value of 4 not until level 10 is reached.

b) The handling code could assume an initial value of 3, and the master skill could could add a value of 0.1 (by Skill.PassivePowerUps) to an attribute like "Additional Number of Triple Skill Arrows".
Regardless of the actual way, the code can then check if the received hits of one animation exceed the number of allowed arrows.

**Comment by sven-n** [comment]
I’m currently working on it and it seems that some more changes will be required.

1. Currently, a master skill just has a definition for level depending damage. Buffs etc. are not yet meant to have level depending effects. It will make more sense to have something general for all use cases.

2. The level depending values are calculated by some crazy formulas. We could either store the resulting values (like we do already for attack damage) or store a formula - I tend to the latter. However, then we need something to evaluate them at runtime, e.g. https://github.com/mariuszgromada/MathParser.org-mXparser

3. If we think a step further, we could even think about making formula evaluations possible within the attribute system infrastructure
